api/src/main/java/org/apache/iceberg/util/WrapperSet.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.util;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterators;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+/**
+ * A custom set for a {@link Wrapper} of the given type that maintains insertion order
+ *
+ * @param <T> The type to wrap in a {@link Wrapper} instance.
+ */
+abstract class WrapperSet<T> implements Set<T>, Serializable {
+  private final Set<Wrapper<T>> set;
+
+  protected WrapperSet(Iterable<Wrapper<T>> wrappers) {
+    this.set = Sets.newLinkedHashSet(wrappers);
+  }
+
+  protected abstract Wrapper<T> wrapper();
+
+  protected abstract Wrapper<T> wrap(T file);
+
+  protected abstract boolean isInstance(Object obj);
```

Comment:
One more alternative is to have `abstract Class<T> elementClass();` to rely on `isInstance` and use various utils like `Iterators.toArray` that accept `Class`. Up to you.

Reply:
yeah this is another place where I wanted to be fully consistent with `CharSequenceSet` which also doesn't throw a  CCE. I think I'd prefer using `elementClass()` here since that gives us slightly more flexibility